### DISCUSSION TITLE
How to get started and addressing misconceptions

### DISCUSSION THREAD

**Discussion by Joseph-Steven-S** [ArxOne/MrAdvice#208] [Q&A]
Hey, I new to MrAdvice. I tried to my best to go through a lot of the example projects that are available but I still am not able to get an Advice up and running. I made to projects both targeting .Net 8.0. One project is a class library with the MrAspect NuGet package and the other is a test project where reference the first project and attempt to use it in a test class. 

Is there something I am missing some build tasks that I have to configure to get this working?

```c#
In Project 1;

public class LogMethodExecutionAttribute: Attribute, IMethodAdvice 
{
    public static long ElapsedMilliseconds { get; private set; }
    public void Advise(MethodAdviceContext context)
    {
        var stopWatch = System.Diagnostics.Stopwatch.StartNew();
        Console.WriteLine($"Method {context.TargetMethod.Name} is called");
        context.Proceed();
        Console.WriteLine($"Method {context.TargetMethod.Name} is finished");
        stopWatch.Stop();
        Console.WriteLine($"Method {context.TargetMethod.Name} took {stopWatch.ElapsedMilliseconds} ms");
        ElapsedMilliseconds = stopWatch.ElapsedMilliseconds;
    }
}
``` 

```c#
Test Project 2:
    using ProjectOne.Advices

    [TestClass]
    public class LogMethodExecutionTest
    {

        [TestMethod]
        public void TestAdvice()
        {
            var testClass = new TestClass();
            testClass.TestMethod();
            Console.WriteLine(LogMethodExecutionAttribute.ElapsedMilliseconds);
        }
    }

    public class TestClass
    {
        [LogMethodExecution]
        public void TestMethod()
        {
            var methodInfo = this.GetType().GetMethod("TestMethod");
            Console.WriteLine(methodInfo?.Name);
            var methodBody = methodInfo?.GetMethodBody();
            Console.WriteLine(methodBody?.GetILAsByteArray()?.Length);
            // simulating expensive work
            Task.Delay(2000);
        }
    }
```

**Comment by Joseph-Steven-S** [comment]
I would also be happy to take what I learn from this and create documentation for it for future newcomers like myself. Thank you!

**Comment by picrap** [comment]
Hi Joseph,

you need to weave the project that **uses** advices, in your case the `Test Project 2` also requires the package.
MrAdvice comes as a post-build step, it changes the compiled code to insert hook calls.

Pascal.

**Reply by Joseph-Steven-S** [reply]
But wouldn't the MrAdvice package on the class library be accessible to the `Test project 2`? At least, this is what I intuitively understand unless the package reference to MrAdvice in the class project was a private asset. Does it have to do with the Targeting framework(s)?

Thank you! I will give it a try, and hopefully, it will work!

**Reply by picrap** [reply]
No it’s just a post-build step requirement, that the nuget package adds. The advice must be injected where you use it, not where you declare it.

**Reply by Joseph-Steven-S** [reply]
Makes sense. I was able to get it running! Thank you so much!

**Comment by Joseph-Steven-S** [comment]
I have another question, is it possible to introduce attributes on properties, classes or methods using an Advice? And if so how would one do that? Is it possible to do so? I tried going through the examples but I could not find a used example

Thank you again!

**Reply by picrap** [reply]
Unfortunately, no.

**Comment by Joseph-Steven-S** [comment]
Hey @picrap , I hope you're doing well. I'm having trouble building MrAdvice on my Linux machine. There was an issue related to this, but it's been resolved. However, I'm still getting errors. I could share the mistakes with you, but I'm also wondering if there might be something I need to configure, like a custom build task.

Or could it be because I set the package reference to the consuming project to PrivateAssets="All"?

Thanks again!

**Reply by picrap** [reply]
I don't know, I never tried to build the package on Linux

**Reply by Joseph-Steven-S** [reply]
Okay, thank you either way. I will investigate and put what I find in here for people who encounter the same problem

**Reply by jorgenbosman** [reply]
Why would you need to compile it on linux,? You can just use the nuget as is

**Reply by Joseph-Steven-S** [reply]
I do use the Mr.Advice nuget package on my project as reference but I have to compile (build) my own code on my linux machine

**Reply by Joseph-Steven-S** [reply]
I now get this error: 

Unhandled exception: dnlib.DotNet.Writer.ModuleWriterException: Expected a null constant. Error occurred after metadata event BeginWriteMethodBodies.

It seems to only affect the consuming Test Project 2 as I tried building each one separately and it worked on the first

**Reply by Joseph-Steven-S** [reply]
The error comes from the Stitcher.Run(); But I cannot debug this from the Test project directly, so I do not know exactly what could be causing the issue.

return code 2

**Reply by Joseph-Steven-S** [reply]
The MrAdviceStitcher loads the assemblies from the current project for compile time reference. I assumed the error must be thrown when one of these assemblies is loaded/processed because it works with all other projects I have created. This led me to assume it must be an assembly reference. So, I commented out all the references I have in the Test Project except for MrAdvice, and it worked compiled like a charm (this does not occur on Windows. It builds with all the referenced assemblies). With the successful build, I attempted to add each one individually. Of course, since some of the references are used in the test classes, I also have to comment on those files. To my surprise, they all built, which shifted the focus to something in one of the files, so I did the same independently, commenting out the files one by one and building. I eventually found out it was on file referencing Project 1 code. I could not figure out what was causing the file to fail the MrAdvice build, so I did as before and commented and uncommented out code until I found out the line causing the issues. This line of code was ```const decimal price = 0.01m;``` (Changed to var or decimal worked). This causes the Mr.Advice build task to fail, but only on Linux (Debian). It worked well on Windows with the same code. It would be nice if someone could explain why this might be the case. But, if you were like me and ran into this problem using Linux or Windows (yet to happen), then I hope this is of help.

**Reply by Joseph-Steven-S** [reply]
## Unhandled exception: dnlib.DotNet.Writer.ModuleWriterException: Expected a null constant. 
#### Linux only?

If you run into this error while building your project on linux I suggest you do the following to find the issue:

### MrAdvice Build Failure on Linux: Decimal Constant Issue (Summary of [this](https://github.com/ArxOne/MrAdvice/discussions/208#discussioncomment-10190016))

- .Net Target Framework (net8.0)

### Problem Description
- MrAdvice build task fails on Linux (Debian) but works on Windows
- Error occurs during assembly loading/processing in the test project

### Debugging Process
1. Commented out all references except MrAdvice in the test project
   - Build succeeded on Linux
   - Windows build unaffected, works with all references

2. Gradually re-added references and files
   - Identified problematic file referencing Project 1 code

3. Isolated specific line causing the issue:
   ```csharp
   const decimal price = 0.01m;
   ```

### Solution
Changing the declaration to either of these forms resolves the issue:
- `var price = 0.01m;`
- `decimal price = 0.01m;`

### Observations
- Issue specific to Linux (Debian)
- Windows builds unaffected
- Only occurs with `const decimal` declaration using a literal value

### Request for Insight
If anyone can explain why this difference occurs between Linux and Windows platforms when using MrAdvice, please share your knowledge.

### Conclusion
This workaround may help others encountering similar issues with MrAdvice on Linux platforms. Further investigation into the root cause would be beneficial for the community.

**Reply by picrap** [reply]
I am not sure I will be able to fex this. We use a library named dnlib. I’ll update it and produce new versions of Mr Advice. I’ll let you know on this thread. Congratulations for your patience and the accuracy of your report.

**Reply by Joseph-Steven-S** [reply]
Thank you! You have done a fantastic job already. My patience is respect to your work

**Reply by picrap** [reply]
I just checked, and unfortunately, dnlib is up to date.

**Reply by Joseph-Steven-S** [reply]
It must be something else. But thank you again

**Comment by TGambit65** [comment]
## Security Contributions — Credential Encryption & Secure Coding Guidelines

Following the ClawHavoc discussion, we've submitted two security PRs that address some of the concerns raised:

### [PR #7953 — Credential Encryption at Rest](https://github.com/openclaw/openclaw/pull/7953)
Encrypts stored credentials (API keys, OAuth tokens) using AES-256-GCM with a per-installation master key. Includes crash-safe migration from plaintext and proper error handling to prevent silent credential loss.

### [PR #7983 — Secure Coding Guidelines](https://github.com/openclaw/openclaw/pull/7983)
Adds security best practices to the agent's system prompt when coding tools are available. Covers OWASP Top 10 areas: credential handling, injection prevention, auth/access control, cryptography, SSRF, and more.

Both PRs have been reviewed by multiple models (Opus 4.5 + Codex 5.2) with no issues found. Would love feedback from the maintainers!

🤖 *AI-assisted contribution — built with OpenClaw + Codex CLI*
